In [1]:
import pandas as pd # A general purpose Python library for data analysis
import numpy as np # A library for scientific computing in Python (e.g., provides high-performance multi-dimensional array objects and operations)

import matplotlib.pyplot as plt # a plotting library for Python and NumPy (readily customizable)
import seaborn as sns # Another plotting library for Python (fewer syntax, excellent default themes, behind the scenes, it uses matplotlib)
import time

## Knowledge Stream Summer 2023

In this notebook, we will learn about the key data structures provided by the Pandas library: **Data Frames, Series, and Indices**.

In addition, we will learn about the following operations:
* How to access data contained in these structures?
* How to read files (e.g., csv, xlsx, sql) to create these structures?
* How to carry out different data manipulation tasks using these structures?

`Dataset`: US elections with information about candidates, their party, votes won, year of election and the result.

## Reading in Data Frames from Files

Pandas has a number of useful file reading tools. You can see them enumerated by typing **"pd.re"** and pressing `tab`. We'll be using **read_csv** today. Note that these file reading functions do all the *data parsing* for you, which is very useful.

Before loading a file into a dataframe, let's first take a look at the **elections.csv** file

In [2]:
elections = pd.read_csv('C:\\Users\\zafar\\Downloads\\KM_notebook\\elections.csv')
elections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

In [8]:
s = pd.Series(['welcome','to', 'data_100','54'])
s

0     welcome
1          to
2    data_100
3          54
dtype: object

In [47]:
s = pd.Series([-1, 10, 2], index = ['a', 'b', 'c2'])
s.values
s.index

Index(['a', 'b', 'c2'], dtype='object')

In [28]:
elections[['Year','Candidate']].values

array([[1824, 'Andrew Jackson'],
       [1824, 'John Quincy Adams'],
       [1828, 'Andrew Jackson'],
       [1828, 'John Quincy Adams'],
       [1832, 'Andrew Jackson'],
       [1832, 'Henry Clay'],
       [1832, 'William Wirt'],
       [1836, 'Hugh Lawson White'],
       [1836, 'Martin Van Buren'],
       [1836, 'William Henry Harrison'],
       [1840, 'Martin Van Buren'],
       [1840, 'William Henry Harrison'],
       [1844, 'Henry Clay'],
       [1844, 'James Polk'],
       [1848, 'Lewis Cass'],
       [1848, 'Martin Van Buren'],
       [1848, 'Zachary Taylor'],
       [1852, 'Franklin Pierce'],
       [1852, 'John P. Hale'],
       [1852, 'Winfield Scott'],
       [1856, 'James Buchanan'],
       [1856, 'John C. Frémont'],
       [1856, 'Millard Fillmore'],
       [1860, 'Abraham Lincoln'],
       [1860, 'John Bell'],
       [1860, 'John C. Breckinridge'],
       [1860, 'Stephen A. Douglas'],
       [1864, 'Abraham Lincoln'],
       [1864, 'George B. McClellan'],
       [1868, 'H

In [54]:
elections[elections['Year'] > 1888]

Year          Candidate                Party  Popular vote Result  \
46   1892  Benjamin Harrison           Republican       5176108   loss   
47   1892   Grover Cleveland           Democratic       5553898    win   
48   1892    James B. Weaver             Populist       1041028   loss   
49   1892       John Bidwell          Prohibition        270879   loss   
50   1896     John M. Palmer  National Democratic        134645   loss   
..    ...                ...                  ...           ...    ...   
177  2016         Jill Stein                Green       1457226   loss   
178  2020       Joseph Biden           Democratic      81268924    win   
179  2020       Donald Trump           Republican      74216154   loss   
180  2020       Jo Jorgensen          Libertarian       1865724   loss   
181  2020     Howard Hawkins                Green        405035   loss   

             %  
46   42.984101  
47   46.121393  
48    8.645038  
49    2.249468  
50    0.969566  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[136 rows x 6 columns]

In [82]:
ele= elections[['Year','Result']].value_counts()
ele.head(60)


Year  Result
1996  loss      6
1948  loss      5
2016  loss      5
1976  loss      5
2004  loss      5
2008  loss      5
1912  loss      4
1904  loss      4
1920  loss      4
1980  loss      4
2000  loss      4
1992  loss      4
1916  loss      3
1908  loss      3
1896  loss      3
1892  loss      3
1932  loss      3
1936  loss      3
1888  loss      3
1988  loss      3
1884  loss      3
1860  loss      3
2012  loss      3
2020  loss      3
1924  loss      2
1836  loss      2
1832  loss      2
1968  loss      2
1900  loss      2
1972  loss      2
1956  loss      2
1928  loss      2
1848  loss      2
1852  loss      2
1952  loss      2
1856  loss      2
1880  loss      2
1984  loss      2
1940  loss      2
1952  win       1
2016  win       1
1956  win       1
1960  loss      1
1948  win       1
1960  win       1
1992  win       1
1964  loss      1
2012  win       1
1964  win       1
1996  win       1
1968  win       1
2004  win       1
1972  win       1
2000  win       1
1976  win      

In [36]:
elections.set_index(('Year' ) ,inplace= True)

In [40]:
elections.head(5)

Candidate                  Party  Popular vote Result          %
Year                                                                          
1824     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1824  John Quincy Adams  Democratic-Republican        113142    win  42.789878
1828     Andrew Jackson             Democratic        642806    win  56.203927
1828  John Quincy Adams    National Republican        500897   loss  43.796073
1832     Andrew Jackson             Democratic        702735    win  54.574789

In [45]:
elections.reset_index( inplace = True)

In [46]:
elections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

In [4]:
pd.DataFrame([[1,'one', 'tp1'],[2, 'two', 'tp3']], columns= ['Number','Description', 'tp'])

Number Description   tp
0       1         one  tp1
1       2         two  tp3

In [88]:
pd.DataFrame({1: ['Number','Number2'], 2: ['Description','Description2']})

1             2
0   Number   Description
1  Number2  Description2

In [89]:
pd.DataFrame([["one","two"], ["four","you"]], columns= [1,2])

1    2
0   one  two
1  four  you

In [93]:
pd.DataFrame({'fruit': ['apple','orange'], 'price': ['234$', '20$']})

fruit price
0   apple  234$
1  orange   20$

In [94]:
elections.columns

Index(['Year', 'Candidate', 'Party', 'Popular vote', 'Result', '%'], dtype='object')

In [95]:
;elections.info

<bound method DataFrame.info of      Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2   

In [126]:
elections.index

RangeIndex(start=0, stop=182, step=1)

In [109]:
elections.shape

(182, 6)

In [123]:
elections[['Year','Candidate']]

Year          Candidate
0    1824     Andrew Jackson
1    1824  John Quincy Adams
2    1828     Andrew Jackson
3    1828  John Quincy Adams
4    1832     Andrew Jackson
..    ...                ...
177  2016         Jill Stein
178  2020       Joseph Biden
179  2020       Donald Trump
180  2020       Jo Jorgensen
181  2020     Howard Hawkins

[182 rows x 2 columns]

In [124]:
elections[elections['Year'] > 1888]

Year          Candidate                Party  Popular vote Result  \
46   1892  Benjamin Harrison           Republican       5176108   loss   
47   1892   Grover Cleveland           Democratic       5553898    win   
48   1892    James B. Weaver             Populist       1041028   loss   
49   1892       John Bidwell          Prohibition        270879   loss   
50   1896     John M. Palmer  National Democratic        134645   loss   
..    ...                ...                  ...           ...    ...   
177  2016         Jill Stein                Green       1457226   loss   
178  2020       Joseph Biden           Democratic      81268924    win   
179  2020       Donald Trump           Republican      74216154   loss   
180  2020       Jo Jorgensen          Libertarian       1865724   loss   
181  2020     Howard Hawkins                Green        405035   loss   

             %  
46   42.984101  
47   46.121393  
48    8.645038  
49    2.249468  
50    0.969566  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[136 rows x 6 columns]

In [ ]:
elections[elections['Year'] > 1888]

In [127]:
elections.loc[[55,78,90],['Candidate', 'Party','Popular vote']]

Candidate         Party  Popular vote
55  William Jennings Bryan    Democratic       6370932
78   Parley P. Christensen  Farmer–Labor        265398
90              Alf Landon    Republican      16679543

In [128]:
elections.loc[0:5 , 'Candidate' : '%']

Candidate                  Party  Popular vote Result          %
0     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1  John Quincy Adams  Democratic-Republican        113142    win  42.789878
2     Andrew Jackson             Democratic        642806    win  56.203927
3  John Quincy Adams    National Republican        500897   loss  43.796073
4     Andrew Jackson             Democratic        702735    win  54.574789
5         Henry Clay    National Republican        484205   loss  37.603628

In [132]:
elections.loc[: , 'Candidate':]

Candidate                  Party  Popular vote Result          %
0       Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1    John Quincy Adams  Democratic-Republican        113142    win  42.789878
2       Andrew Jackson             Democratic        642806    win  56.203927
3    John Quincy Adams    National Republican        500897   loss  43.796073
4       Andrew Jackson             Democratic        702735    win  54.574789
..                 ...                    ...           ...    ...        ...
177         Jill Stein                  Green       1457226   loss   1.073699
178       Joseph Biden             Democratic      81268924    win  51.311515
179       Donald Trump             Republican      74216154   loss  46.858542
180       Jo Jorgensen            Libertarian       1865724   loss   1.177979
181     Howard Hawkins                  Green        405035   loss   0.255731

[182 rows x 5 columns]

In [133]:
elections.iloc[1:3 , 2:4]

Party  Popular vote
1  Democratic-Republican        113142
2             Democratic        642806

In [134]:
elections.iloc[[1,3,4] , [2,4]]

Party Result
1  Democratic-Republican    win
3    National Republican   loss
4             Democratic    win

In [135]:
elections.iloc[1:3 , 0:4]

Year          Candidate                  Party  Popular vote
1  1824  John Quincy Adams  Democratic-Republican        113142
2  1828     Andrew Jackson             Democratic        642806

In [136]:
elections.iloc[: , 0:4]

Year          Candidate                  Party  Popular vote
0    1824     Andrew Jackson  Democratic-Republican        151271
1    1824  John Quincy Adams  Democratic-Republican        113142
2    1828     Andrew Jackson             Democratic        642806
3    1828  John Quincy Adams    National Republican        500897
4    1832     Andrew Jackson             Democratic        702735
..    ...                ...                    ...           ...
177  2016         Jill Stein                  Green       1457226
178  2020       Joseph Biden             Democratic      81268924
179  2020       Donald Trump             Republican      74216154
180  2020       Jo Jorgensen            Libertarian       1865724
181  2020     Howard Hawkins                  Green        405035

[182 rows x 4 columns]

In [140]:
elections.iloc[[1,2,3] , 0]

1    1824
2    1828
3    1828
Name: Year, dtype: int64

In [151]:
elections.iloc[1:3 , 1::3]

Candidate Result
1  John Quincy Adams    win
2     Andrew Jackson    win

In [162]:
elections[['Candidate']]

Candidate
0       Andrew Jackson
1    John Quincy Adams
2       Andrew Jackson
3    John Quincy Adams
4       Andrew Jackson
..                 ...
177         Jill Stein
178       Joseph Biden
179       Donald Trump
180       Jo Jorgensen
181     Howard Hawkins

[182 rows x 1 columns]

In [168]:
elections[1:30 :2]

Year               Candidate                  Party  Popular vote Result  \
1   1824       John Quincy Adams  Democratic-Republican        113142    win   
3   1828       John Quincy Adams    National Republican        500897   loss   
5   1832              Henry Clay    National Republican        484205   loss   
7   1836       Hugh Lawson White                   Whig        146109   loss   
9   1836  William Henry Harrison                   Whig        550816   loss   
11  1840  William Henry Harrison                   Whig       1275583    win   
13  1844              James Polk             Democratic       1339570    win   
15  1848        Martin Van Buren              Free Soil        291501   loss   
17  1852         Franklin Pierce             Democratic       1605943    win   
19  1852          Winfield Scott                   Whig       1386942   loss   
21  1856         John C. Frémont             Republican       1342345   loss   
23  1860         Abraham Lincoln             Republican       1855993    win   
25  1860    John C. Breckinridge    Southern Democratic        848019   loss   
27  1864         Abraham Lincoln         National Union       2211317    win   
29  1868         Horatio Seymour             Democratic       2708744   loss   

            %  
1   42.789878  
3   43.796073  
5   37.603628  
7   10.005985  
9   37.721543  
11  53.051213  
13  50.749477  
15  10.138474  
17  51.013168  
19  44.056548  
21  33.139919  
23  39.699408  
25  18.138998  
27  54.951512  
29  47.334695

In [172]:
elections.loc[0:9 , :]

Year               Candidate                  Party  Popular vote Result  \
0  1824          Andrew Jackson  Democratic-Republican        151271   loss   
1  1824       John Quincy Adams  Democratic-Republican        113142    win   
2  1828          Andrew Jackson             Democratic        642806    win   
3  1828       John Quincy Adams    National Republican        500897   loss   
4  1832          Andrew Jackson             Democratic        702735    win   
5  1832              Henry Clay    National Republican        484205   loss   
6  1832            William Wirt           Anti-Masonic        100715   loss   
7  1836       Hugh Lawson White                   Whig        146109   loss   
8  1836        Martin Van Buren             Democratic        763291    win   
9  1836  William Henry Harrison                   Whig        550816   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628  
6   7.821583  
7  10.005985  
8  52.272472  
9  37.721543

In [181]:
elections[(elections['Result'] == 'loss')]

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
3    1828  John Quincy Adams    National Republican        500897   loss   
5    1832         Henry Clay    National Republican        484205   loss   
6    1832       William Wirt           Anti-Masonic        100715   loss   
7    1836  Hugh Lawson White                   Whig        146109   loss   
..    ...                ...                    ...           ...    ...   
176  2016    Hillary Clinton             Democratic      65853514   loss   
177  2016         Jill Stein                  Green       1457226   loss   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
3    43.796073  
5    37.603628  
6     7.821583  
7    10.005985  
..         ...  
176  48.521539  
177   1.073699  
179  46.858542  
180   1.177979  
181   0.255731  

[132 rows x 6 columns]

In [191]:
elections[(elections['Result'] == 'loss') & (elections['Year'] > 2000)]

Year         Candidate         Party  Popular vote Result          %
156  2004        David Cobb         Green        119859   loss   0.098088
158  2004        John Kerry    Democratic      59028444   loss  48.306775
159  2004  Michael Badnarik   Libertarian        397265   loss   0.325108
160  2004  Michael Peroutka  Constitution        143630   loss   0.117542
161  2004       Ralph Nader   Independent        465151   loss   0.380663
163  2008          Bob Barr   Libertarian        523715   loss   0.399565
164  2008     Chuck Baldwin  Constitution        199750   loss   0.152398
165  2008  Cynthia McKinney         Green        161797   loss   0.123442
166  2008       John McCain    Republican      59948323   loss  45.737243
167  2008       Ralph Nader   Independent        739034   loss   0.563842
169  2012      Gary Johnson   Libertarian       1275971   loss   0.992241
170  2012        Jill Stein         Green        469627   loss   0.365199
171  2012       Mitt Romney    Republican      60933504   loss  47.384076
172  2016    Darrell Castle  Constitution        203091   loss   0.149640
174  2016     Evan McMullin   Independent        732273   loss   0.539546
175  2016      Gary Johnson   Libertarian       4489235   loss   3.307714
176  2016   Hillary Clinton    Democratic      65853514   loss  48.521539
177  2016        Jill Stein         Green       1457226   loss   1.073699
179  2020      Donald Trump    Republican      74216154   loss  46.858542
180  2020      Jo Jorgensen   Libertarian       1865724   loss   1.177979
181  2020    Howard Hawkins         Green        405035   loss   0.255731

In [192]:
elections[(elections['Result'] == 'win') & (elections['Year'] > 2000)]

Year       Candidate       Party  Popular vote Result          %
157  2004  George W. Bush  Republican      62040610    win  50.771824
162  2008    Barack Obama  Democratic      69498516    win  53.023510
168  2012    Barack Obama  Democratic      65915795    win  51.258484
173  2016    Donald Trump  Republican      62984828    win  46.407862
178  2020    Joseph Biden  Democratic      81268924    win  51.311515

In [195]:
candidat = ['George W. Bush', 'Donald Trump']

elections[elections['Candidate'].isin(candidat)]

Year       Candidate       Party  Popular vote Result          %
152  2000  George W. Bush  Republican      50456002    win  47.974666
157  2004  George W. Bush  Republican      62040610    win  50.771824
173  2016    Donald Trump  Republican      62984828    win  46.407862
179  2020    Donald Trump  Republican      74216154   loss  46.858542

In [198]:
elections[elections['Candidate'].str.startswith('D')]

Year          Candidate         Party  Popular vote Result          %
106  1952  Dwight Eisenhower    Republican      34075529    win  55.325173
109  1956  Dwight Eisenhower    Republican      35579180    win  57.650654
132  1984     David Bergland   Libertarian        228111   loss   0.247245
156  2004         David Cobb         Green        119859   loss   0.098088
172  2016     Darrell Castle  Constitution        203091   loss   0.149640
173  2016       Donald Trump    Republican      62984828    win  46.407862
179  2020       Donald Trump    Republican      74216154   loss  46.858542

In [205]:
elections[elections['Candidate'].str.endswith('nson')]

Year        Candidate        Party  Popular vote Result          %
71   1916  Allan L. Benson    Socialist        590524   loss   3.194193
105  1952  Adlai Stevenson   Democratic      27375090   loss  44.446312
108  1956  Adlai Stevenson   Democratic      26028028   loss  42.174464
114  1964   Lyndon Johnson   Democratic      43127041    win  61.344703
169  2012     Gary Johnson  Libertarian       1275971   loss   0.992241
175  2016     Gary Johnson  Libertarian       4489235   loss   3.307714

In [206]:
elections[elections['Party']== 'Socialist']

Year        Candidate      Party  Popular vote Result         %
58   1904   Eugene V. Debs  Socialist        402810   loss  2.985897
62   1908   Eugene V. Debs  Socialist        420852   loss  2.850866
66   1912   Eugene V. Debs  Socialist        901551   loss  6.004354
71   1916  Allan L. Benson  Socialist        590524   loss  3.194193
76   1920   Eugene V. Debs  Socialist        913693   loss  3.428282
85   1928    Norman Thomas  Socialist        267478   loss  0.728623
88   1932    Norman Thomas  Socialist        884885   loss  2.236211
92   1936    Norman Thomas  Socialist        187910   loss  0.412876
95   1940    Norman Thomas  Socialist        116599   loss  0.234237
102  1948    Norman Thomas  Socialist        139569   loss  0.286312

In [ ]:
#adding cloum

In [247]:
election = elections['Candidate'].value_counts()
elections['Candidate_count'] = election


In [248]:
elections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  Candidate_count  
0    57.210122              NaN  
1    42.789878              NaN  
2    56.203927              NaN  
3    43.796073              NaN  
4    54.574789              NaN  
..         ...              ...  
177   1.073699              NaN  
178  51.311515              NaN  
179  46.858542              NaN  
180   1.177979              NaN  
181   0.255731              NaN  

[182 rows x 7 columns]

In [246]:
elections

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  Candidate_count  
0    57.210122              NaN  
1    42.789878              NaN  
2    56.203927              NaN  
3    43.796073              NaN  
4    54.574789              NaN  
..         ...              ...  
177   1.073699              NaN  
178  51.311515              NaN  
179  46.858542              NaN  
180   1.177979              NaN  
181   0.255731              NaN  

[182 rows x 7 columns]

In [255]:
elections.loc[elections['Candidate'] == 'Donald Trump', 'Party']

173    Republican
179    Republican
Name: Party, dtype: object

In [264]:
elections[elections['Year'] == 2000].sample(2)

Year       Candidate        Party  Popular vote Result          %  \
152  2000  George W. Bush   Republican      50456002    win  47.974666   
153  2000    Harry Browne  Libertarian        384431   loss   0.365525   

     Candidate_count  
152              NaN  
153              NaN

In [268]:
elections['Candidate'].sort_values()

75           Aaron S. Watkins
27            Abraham Lincoln
23            Abraham Lincoln
108           Adlai Stevenson
105           Adlai Stevenson
                ...          
19             Winfield Scott
37     Winfield Scott Hancock
74             Woodrow Wilson
70             Woodrow Wilson
16             Zachary Taylor
Name: Candidate, Length: 182, dtype: object

In [275]:
elections['Candidate'].sort_values()

75           Aaron S. Watkins
27            Abraham Lincoln
23            Abraham Lincoln
108           Adlai Stevenson
105           Adlai Stevenson
                ...          
19             Winfield Scott
37     Winfield Scott Hancock
74             Woodrow Wilson
70             Woodrow Wilson
16             Zachary Taylor
Name: Candidate, Length: 182, dtype: object

In [ ]:
#using lamda funcation create column and sort value without crating ne w column map function

In [294]:
elections.groupby("Party")[['Year']].agg(sum)

Year
Party                       
American                3832
American Independent    5916
Anti-Masonic            1832
Anti-Monopoly           1884
Citizens                1980
Communist               1932
Constitution            6028
Constitutional Union    1860
Democratic             90544
Democratic-Republican   3648
Dixiecrat               1948
Farmer–Labor            1920
Free Soil               3700
Green                  14056
Greenback               1880
Independent            11976
Liberal Republican      1872
Libertarian            23976
National Democratic     1896
National Republican     3660
National Union          1864
Natural Law             1996
New Alliance            1988
Northern Democratic     1860
Populist                5788
Progressive             7736
Prohibition            20968
Reform                  3996
Republican             79532
Socialist              19244
Southern Democratic     1860
States' Rights          1956
Taxpayers               1996
Union                   1936
Union Labor             1888
Whig                   11056

In [25]:
datt.set_index(['a','b','c','d'])

KeyError: "None of ['a', 'b', 'c', 'd'] are in the columns"

In [27]:
datt= pd.DataFrame([['cat', 2.5, 1,'yes'],['cat', 3, 3, 'yes'],['snake', 3, 3, 'yes']
             ,['dog', 5, 'nan', 'yes']], columns= ['animal','age', 'visits','priority'], index=['a', 'b', 'c', 'd'])

datt

animal  age visits priority
a    cat  2.5      1      yes
b    cat  3.0      3      yes
c  snake  3.0      3      yes
d    dog  5.0    nan      yes

In [103]:
data = pd.DataFrame({'animal': ['cat', 'cat', 'snake', 'dog', 'dog', 'cat', 'snake', 'cat', 'dog', 'dog'], 

             'age': [2.5, 3, 0.5, np.nan, 5, 2, 4.5, np.nan, 7, 3],

             'visits': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],

             'priority': ['yes', 'yes', 'no', 'yes', 'no', 'no', 'no', 'yes', 'no', 'no']},

index = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j'])

In [104]:
data

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [37]:
data.info

<bound method DataFrame.info of   animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no>

In [36]:
data.describe

<bound method NDFrame.describe of   animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c  snake  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g  snake  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no>

In [38]:
data.tail(3)

animal  age  visits priority
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [49]:
data.iloc[-3:]

animal  age  visits priority
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [58]:
data[['animal', 'age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [65]:
data.iloc[:, 0:2]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [67]:
data.loc[:, ['animal','age']]

animal  age
a    cat  2.5
b    cat  3.0
c  snake  0.5
d    dog  NaN
e    dog  5.0
f    cat  2.0
g  snake  4.5
h    cat  NaN
i    dog  7.0
j    dog  3.0

In [71]:
data[data['visits']>=3]

animal  age  visits priority
b    cat  3.0       3      yes
d    dog  NaN       3      yes
f    cat  2.0       3       no

In [77]:
data['animal'].value_counts()

cat      4
dog      4
snake    2
Name: animal, dtype: int64

In [105]:
data['animal'] = data['animal'].replace('snake','pyt')
data


animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c    pyt  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g    pyt  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [106]:
data

animal  age  visits priority
a    cat  2.5       1      yes
b    cat  3.0       3      yes
c    pyt  0.5       2       no
d    dog  NaN       3      yes
e    dog  5.0       2       no
f    cat  2.0       3       no
g    pyt  4.5       1       no
h    cat  NaN       1      yes
i    dog  7.0       2       no
j    dog  3.0       1       no

In [121]:
data.sort_values(by = ['age','visits'] , ascending = [True , False])

animal  age  visits priority
c    pyt  0.5       2       no
f    cat  2.0       3       no
a    cat  2.5       1      yes
b    cat  3.0       3      yes
j    dog  3.0       1       no
g    pyt  4.5       1       no
e    dog  5.0       2       no
i    dog  7.0       2       no
d    dog  NaN       3      yes
h    cat  NaN       1      yes

In [125]:
data[data['age'].isnull()]

animal  age  visits priority
d    dog  NaN       3      yes
h    cat  NaN       1      yes

In [129]:
datt.loc[['a','c'],['age','visits']]

age visits
a  2.5      1
c  3.0      3

We can use the **head command** to show only a few rows of a dataframe.

# heading
## heading2

In [7]:
elections.head(10)

Year               Candidate                  Party  Popular vote Result  \
0  1824          Andrew Jackson  Democratic-Republican        151271   loss   
1  1824       John Quincy Adams  Democratic-Republican        113142    win   
2  1828          Andrew Jackson             Democratic        642806    win   
3  1828       John Quincy Adams    National Republican        500897   loss   
4  1832          Andrew Jackson             Democratic        702735    win   
5  1832              Henry Clay    National Republican        484205   loss   
6  1832            William Wirt           Anti-Masonic        100715   loss   
7  1836       Hugh Lawson White                   Whig        146109   loss   
8  1836        Martin Van Buren             Democratic        763291    win   
9  1836  William Henry Harrison                   Whig        550816   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628  
6   7.821583  
7  10.005985  
8  52.272472  
9  37.721543

There is also a **tail command**.

In [9]:
elections.tail(5)

Year       Candidate        Party  Popular vote Result          %
177  2016      Jill Stein        Green       1457226   loss   1.073699
178  2020    Joseph Biden   Democratic      81268924    win  51.311515
179  2020    Donald Trump   Republican      74216154   loss  46.858542
180  2020    Jo Jorgensen  Libertarian       1865724   loss   1.177979
181  2020  Howard Hawkins        Green        405035   loss   0.255731

The `read_csv` command lets us specify a **column to use an index**. For example, we could have used __Year__ as the index.

In [10]:
elections.set_index('Year')

Candidate                  Party  Popular vote Result          %
Year                                                                          
1824     Andrew Jackson  Democratic-Republican        151271   loss  57.210122
1824  John Quincy Adams  Democratic-Republican        113142    win  42.789878
1828     Andrew Jackson             Democratic        642806    win  56.203927
1828  John Quincy Adams    National Republican        500897   loss  43.796073
1832     Andrew Jackson             Democratic        702735    win  54.574789
...                 ...                    ...           ...    ...        ...
2016         Jill Stein                  Green       1457226   loss   1.073699
2020       Joseph Biden             Democratic      81268924    win  51.311515
2020       Donald Trump             Republican      74216154   loss  46.858542
2020       Jo Jorgensen            Libertarian       1865724   loss   1.177979
2020     Howard Hawkins                  Green        405035   loss   0.255731

[182 rows x 5 columns]

Alternately, we could have used the **set_index** commmand on the dataframe.

In [11]:
elections.reset_index()

index  Year          Candidate                  Party  Popular vote  \
0        0  1824     Andrew Jackson  Democratic-Republican        151271   
1        1  1824  John Quincy Adams  Democratic-Republican        113142   
2        2  1828     Andrew Jackson             Democratic        642806   
3        3  1828  John Quincy Adams    National Republican        500897   
4        4  1832     Andrew Jackson             Democratic        702735   
..     ...   ...                ...                    ...           ...   
177    177  2016         Jill Stein                  Green       1457226   
178    178  2020       Joseph Biden             Democratic      81268924   
179    179  2020       Donald Trump             Republican      74216154   
180    180  2020       Jo Jorgensen            Libertarian       1865724   
181    181  2020     Howard Hawkins                  Green        405035   

    Result          %  
0     loss  57.210122  
1      win  42.789878  
2      win  56.203927  
3     loss  43.796073  
4      win  54.574789  
..     ...        ...  
177   loss   1.073699  
178    win  51.311515  
179   loss  46.858542  
180   loss   1.177979  
181   loss   0.255731  

[182 rows x 7 columns]

# Caution:
The **set_index command** (along with all other data frame methods) **does not modify the dataframe**, i.e., the original "elections" is untouched. Note: There is a flag called "inplace" which does modify the calling dataframe (e.g., `elections.set_index("Party",inplace=True)`).

## Duplicate Columns?
By contast, column names MUST be unique. For example, if we try to read in a file for which column names are not unique, Pandas will automatically any duplicates. Load duplicate_columns.csv

In [12]:
duplicate = pd.read_csv('C:\\Users\\zafar\\Downloads\\duplicate_columns.csv')
duplicate

name    name.1      flavor
0   john     smith     vanilla
1  zhang      shan   chocolate
2  fulan  alfulani  strawberry
3   hong   gildong      banana

## The [ ] Operator & Indexing

The DataFrame class has an indexing operator **[ ]** (also known as the 'brack' operator) that lets you do a variety of different things. If your provide a String to the **[ ]** operator, you get back a ***Series*** corresponding to the requested label.

1.Use **[ ]** to display different columns

2.Use List retrive multiple columns

In [14]:
# Answer Here
candiate = elections["Candidate"].value_counts().head(5)
print(candiate)
multip_col = elections.loc[:,['Candidate','Party','Year']]
print(multip_col)

Norman Thomas         5
Ralph Nader           4
Franklin Roosevelt    4
Eugene V. Debs        4
Andrew Jackson        3
Name: Candidate, dtype: int64
             Candidate                  Party  Year
0       Andrew Jackson  Democratic-Republican  1824
1    John Quincy Adams  Democratic-Republican  1824
2       Andrew Jackson             Democratic  1828
3    John Quincy Adams    National Republican  1828
4       Andrew Jackson             Democratic  1832
..                 ...                    ...   ...
177         Jill Stein                  Green  2016
178       Joseph Biden             Democratic  2020
179       Donald Trump             Republican  2020
180       Jo Jorgensen            Libertarian  2020
181     Howard Hawkins                  Green  2020

[182 rows x 3 columns]


The **[ ]** operator also accepts a list of strings. In this case, you get back a **DataFrame** corresponding to the requested strings.

In [16]:
elections[3:20]

Year               Candidate                Party  Popular vote Result  \
3   1828       John Quincy Adams  National Republican        500897   loss   
4   1832          Andrew Jackson           Democratic        702735    win   
5   1832              Henry Clay  National Republican        484205   loss   
6   1832            William Wirt         Anti-Masonic        100715   loss   
7   1836       Hugh Lawson White                 Whig        146109   loss   
8   1836        Martin Van Buren           Democratic        763291    win   
9   1836  William Henry Harrison                 Whig        550816   loss   
10  1840        Martin Van Buren           Democratic       1128854   loss   
11  1840  William Henry Harrison                 Whig       1275583    win   
12  1844              Henry Clay                 Whig       1300004   loss   
13  1844              James Polk           Democratic       1339570    win   
14  1848              Lewis Cass           Democratic       1223460   loss   
15  1848        Martin Van Buren            Free Soil        291501   loss   
16  1848          Zachary Taylor                 Whig       1360235    win   
17  1852         Franklin Pierce           Democratic       1605943    win   
18  1852            John P. Hale            Free Soil        155210   loss   
19  1852          Winfield Scott                 Whig       1386942   loss   

            %  
3   43.796073  
4   54.574789  
5   37.603628  
6    7.821583  
7   10.005985  
8   52.272472  
9   37.721543  
10  46.948787  
11  53.051213  
12  49.250523  
13  50.749477  
14  42.552229  
15  10.138474  
16  47.309296  
17  51.013168  
18   4.930283  
19  44.056548

A list of one label also returns a DataFrame. This can be handy if you want your results as a DataFrame, not a series.

Note that we can also use the **to_frame** method to turn a Series into a DataFrame.

Extract one col name "Candidates" from DataFrame it will be a series. Convert series into a DataFrame.

In [18]:
elections["Year"].to_frame()

Year
0    1824
1    1824
2    1828
3    1828
4    1832
..    ...
177  2016
178  2020
179  2020
180  2020
181  2020

[182 rows x 1 columns]

In [ ]:
# Answer Here

### Row Indexing

The `[]` operator also accepts numerical slices as arguments. In this case, we are indexing by row, not column!

Extract few rows from DataFrame

In [20]:
# Answer Here
elections[0:10]

Year               Candidate                  Party  Popular vote Result  \
0  1824          Andrew Jackson  Democratic-Republican        151271   loss   
1  1824       John Quincy Adams  Democratic-Republican        113142    win   
2  1828          Andrew Jackson             Democratic        642806    win   
3  1828       John Quincy Adams    National Republican        500897   loss   
4  1832          Andrew Jackson             Democratic        702735    win   
5  1832              Henry Clay    National Republican        484205   loss   
6  1832            William Wirt           Anti-Masonic        100715   loss   
7  1836       Hugh Lawson White                   Whig        146109   loss   
8  1836        Martin Van Buren             Democratic        763291    win   
9  1836  William Henry Harrison                   Whig        550816   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628  
6   7.821583  
7  10.005985  
8  52.272472  
9  37.721543

If you provide a single argument to the `[]` operator, it tries to use it as a name. This is true even if the argument passed to **[ ]** is an integer.

In [21]:
elections[0] #this does not work, try uncommenting this to see it fail in action, woo

KeyError: 0

The following cells allow you to **test your understanding**. Let's go over the summary of what we have learnt (see slides).

# Creating DataFrames
Create DataFrame using List and Columns name.

In [22]:
data = [
    ['Toyota', 'Corolla', 2020, 30000],['Honda', 'Civic', 2019, 27000],['Ford', 'Focus', 2018, 25000],
    ['Chevrolet', 'Malibu', 2021, 32000],['Nissan', 'Sentra', 2020, 29000]]

# Col
columns = ['Brand', 'Model', 'Year', 'Mileage']

car_df = pd.DataFrame(data, columns=columns)
print(car_df)


       Brand    Model  Year  Mileage
0     Toyota  Corolla  2020    30000
1      Honda    Civic  2019    27000
2       Ford    Focus  2018    25000
3  Chevrolet   Malibu  2021    32000
4     Nissan   Sentra  2020    29000


Creating DataFrames using **Dictionary**.

In [23]:
# Answer Here

#car data
data = {
    'Brand': ['Toyota', 'Honda', 'Ford', 'Chevrolet', 'Nissan'],
    'Model': ['Corolla', 'Civic', 'Focus', 'Malibu', 'Sentra'],
    'Year': [2020, 2019, 2018, 2021, 2020],
    'Mileage': [30000, 27000, 25000, 32000, 29000]
}

car_df = pd.DataFrame(data)
print(car_df)


       Brand    Model  Year  Mileage
0     Toyota  Corolla  2020    30000
1      Honda    Civic  2019    27000
2       Ford    Focus  2018    25000
3  Chevrolet   Malibu  2021    32000
4     Nissan   Sentra  2020    29000


## Filtering via Boolean Array Selection

The `[]` operator also supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows. The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

In [ ]:
elections[[False, False, False, False, False,
          False, False, True, False, False,
          True, False, False, False, True,
          False, False, False, False, False,
          False, True, False]]

One very common task in Data Science is **filtering**. Boolean Array Selection is one way to achieve this in Pandas. We start by observing that **logical operators** like the equality operator can be applied to **Pandas Series data** to generate a **Boolean Array**.

Compare the 'Result' column to the String 'win' and Show results

In [24]:
#Answer Here
elections['Result'] == 'win'

0      False
1       True
2       True
3      False
4       True
       ...  
177    False
178     True
179    False
180    False
181    False
Name: Result, Length: 182, dtype: bool

Compare the 'Party' column to the String 'Democratic' and Show results

In [25]:
#Answer Here
pd.set_option('display.max_rows',None) 
elections["Party"]=="Democratic"

0      False
1      False
2       True
3      False
4       True
5      False
6      False
7      False
8       True
9      False
10      True
11     False
12     False
13      True
14      True
15     False
16     False
17      True
18     False
19     False
20      True
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28      True
29      True
30     False
31     False
32     False
33     False
34      True
35     False
36     False
37      True
38     False
39      True
40     False
41     False
42     False
43     False
44     False
45      True
46     False
47      True
48     False
49     False
50     False
51     False
52      True
53     False
54     False
55      True
56     False
57      True
58     False
59     False
60     False
61     False
62     False
63     False
64      True
65     False
66     False
67     False
68     False
69     False
70      True
71     False
72     False
73     False
74      True
75     False
76     False

The output of the logical operator applied to the Series is **another Series with the same name and index, but of datatype boolean**.

These boolean Series can be used as an argument to the `[]` operator.

Creates  DataFrame of all election winners since 1980.

In [27]:
#Answer Here
election_Winner = elections['Result'] == 'win'
sin_1980 = elections['Year'] >= 1980

comb = election_Winner & sin_1980

elections[comb]

Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666
157  2004     George W. Bush  Republican      62040610    win  50.771824
162  2008       Barack Obama  Democratic      69498516    win  53.023510
168  2012       Barack Obama  Democratic      65915795    win  51.258484
173  2016       Donald Trump  Republican      62984828    win  46.407862
178  2020       Joseph Biden  Democratic      81268924    win  51.311515

Above, we've assigned the result of the logical operator to a new variable called `iswin`. This is uncommon. Usually, the series is created and used on the same line. Such code is a little tricky to read at first, but you'll get used to it quickly.

Show all 'win' results between 1980 to 2000

In [28]:
from1980to2000 = (elections['Year'] >= 1980) & (elections['Year'] <= 2000)
combi = election_Winner & from1980to2000
elections[combi]

Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666

Show all 'Loss' results of Independent party

In [29]:
ind_Party = elections["Party"] == "Independent"
lossp = elections['Result'] == 'loss'

comb_loss = ind_Party & lossp
elections[comb_loss]



Year         Candidate        Party  Popular vote Result          %
121  1976   Eugene McCarthy  Independent        740460   loss   0.911649
130  1980  John B. Anderson  Independent       5719850   loss   6.631143
143  1992        Ross Perot  Independent      19743821   loss  18.956298
161  2004       Ralph Nader  Independent        465151   loss   0.380663
167  2008       Ralph Nader  Independent        739034   loss   0.563842
174  2016     Evan McMullin  Independent        732273   loss   0.539546

We can select multiple criteria by creating multiple boolean Series and combining them using the `&` operator.

Show results of win with percentage less than 50%

In [30]:
less50 = elections['%'] < 50
elections[less50]

Year               Candidate                  Party  Popular vote Result  \
1    1824       John Quincy Adams  Democratic-Republican        113142    win   
3    1828       John Quincy Adams    National Republican        500897   loss   
5    1832              Henry Clay    National Republican        484205   loss   
6    1832            William Wirt           Anti-Masonic        100715   loss   
7    1836       Hugh Lawson White                   Whig        146109   loss   
9    1836  William Henry Harrison                   Whig        550816   loss   
10   1840        Martin Van Buren             Democratic       1128854   loss   
12   1844              Henry Clay                   Whig       1300004   loss   
14   1848              Lewis Cass             Democratic       1223460   loss   
15   1848        Martin Van Buren              Free Soil        291501   loss   
16   1848          Zachary Taylor                   Whig       1360235    win   
18   1852            John P. Hale              Free Soil        155210   loss   
19   1852          Winfield Scott                   Whig       1386942   loss   
20   1856          James Buchanan             Democratic       1835140    win   
21   1856         John C. Frémont             Republican       1342345   loss   
22   1856        Millard Fillmore               American        873053   loss   
23   1860         Abraham Lincoln             Republican       1855993    win   
24   1860               John Bell   Constitutional Union        590901   loss   
25   1860    John C. Breckinridge    Southern Democratic        848019   loss   
26   1860      Stephen A. Douglas    Northern Democratic       1380202   loss   
28   1864     George B. McClellan             Democratic       1812807   loss   
29   1868         Horatio Seymour             Democratic       2708744   loss   
31   1872          Horace Greeley     Liberal Republican       2834761   loss   
33   1876        Rutherford Hayes             Republican       4034142    win   
35   1880         James B. Weaver              Greenback        308649   loss   
36   1880          James Garfield             Republican       4453337    win   
37   1880  Winfield Scott Hancock             Democratic       4444976   loss   
38   1884         Benjamin Butler          Anti-Monopoly        134294   loss   
39   1884        Grover Cleveland             Democratic       4914482    win   
40   1884         James G. Blaine             Republican       4856905   loss   
41   1884           John St. John            Prohibition        147482   loss   
42   1888          Alson Streeter            Union Labor        146602   loss   
43   1888       Benjamin Harrison             Republican       5443633    win   
44   1888         Clinton B. Fisk            Prohibition        249819   loss   
45   1888        Grover Cleveland             Democratic       5534488   loss   
46   1892       Benjamin Harrison             Republican       5176108   loss   
47   1892        Grover Cleveland             Democratic       5553898    win   
48   1892         James B. Weaver               Populist       1041028   loss   
49   1892            John Bidwell            Prohibition        270879   loss   
50   1896          John M. Palmer    National Democratic        134645   loss   
51   1896         Joshua Levering            Prohibition        131312   loss   
52   1896  William Jennings Bryan             Democratic       6509052   loss   
54   1900         John G. Woolley            Prohibition        210864   loss   
55   1900  William Jennings Bryan             Democratic       6370932   loss   
57   1904         Alton B. Parker             Democratic       5083880   loss   
58   1904          Eugene V. Debs              Socialist        402810   loss   
59   1904        Silas C. Swallow            Prohibition        259102   loss   
61   1904        Thomas E. Watson               Populist        114070   loss   
62   1908          Eugene V. Debs   

Show all 'win' results between 1980 to 2000

In [32]:
# Answer Here
f1980to2000 = (elections['Year'] >= 1980) & (elections['Year'] <= 2000)
combine = election_Winner & f1980to2000  
elections[combine]

Year          Candidate       Party  Popular vote Result          %
131  1980      Ronald Reagan  Republican      43903230    win  50.897944
133  1984      Ronald Reagan  Republican      54455472    win  59.023326
135  1988  George H. W. Bush  Republican      48886597    win  53.518845
140  1992       Bill Clinton  Democratic      44909806    win  43.118485
144  1996       Bill Clinton  Democratic      47400125    win  49.296938
152  2000     George W. Bush  Republican      50456002    win  47.974666

## Loc and iLoc

Show 5 enteries from start

In [33]:
# Answer Here
elections.loc[:5]

Year          Candidate                  Party  Popular vote Result  \
0  1824     Andrew Jackson  Democratic-Republican        151271   loss   
1  1824  John Quincy Adams  Democratic-Republican        113142    win   
2  1828     Andrew Jackson             Democratic        642806    win   
3  1828  John Quincy Adams    National Republican        500897   loss   
4  1832     Andrew Jackson             Democratic        702735    win   
5  1832         Henry Clay    National Republican        484205   loss   

           %  
0  57.210122  
1  42.789878  
2  56.203927  
3  43.796073  
4  54.574789  
5  37.603628

You can provide `.loc` a list of row labels [0-5] and column labels ['Candidate','Party', 'Year'] as input to return a dataframe

In [34]:
# Answer Here
elections.loc[0:5,["Candidate","Party","Year"]]

Candidate                  Party  Year
0     Andrew Jackson  Democratic-Republican  1824
1  John Quincy Adams  Democratic-Republican  1824
2     Andrew Jackson             Democratic  1828
3  John Quincy Adams    National Republican  1828
4     Andrew Jackson             Democratic  1832
5         Henry Clay    National Republican  1832

Loc also supports **slicing** (for all types, including numeric and string labels!). Note that the slicing for loc is **inclusive**, even for numeric slices.

Use Slicing on Rows and Columns

In [35]:
# Answer Here
elections.loc[0:10, ['Candidate',"Year","Result"]]

Candidate  Year Result
0           Andrew Jackson  1824   loss
1        John Quincy Adams  1824    win
2           Andrew Jackson  1828    win
3        John Quincy Adams  1828   loss
4           Andrew Jackson  1832    win
5               Henry Clay  1832   loss
6             William Wirt  1832   loss
7        Hugh Lawson White  1836   loss
8         Martin Van Buren  1836    win
9   William Henry Harrison  1836   loss
10        Martin Van Buren  1840   loss

If we provide only a **single label** for the column argument, we get back a **Series**.

In [36]:
# Answer Here
elections['Year']

0      1824
1      1824
2      1828
3      1828
4      1832
5      1832
6      1832
7      1836
8      1836
9      1836
10     1840
11     1840
12     1844
13     1844
14     1848
15     1848
16     1848
17     1852
18     1852
19     1852
20     1856
21     1856
22     1856
23     1860
24     1860
25     1860
26     1860
27     1864
28     1864
29     1868
30     1868
31     1872
32     1872
33     1876
34     1876
35     1880
36     1880
37     1880
38     1884
39     1884
40     1884
41     1884
42     1888
43     1888
44     1888
45     1888
46     1892
47     1892
48     1892
49     1892
50     1896
51     1896
52     1896
53     1896
54     1900
55     1900
56     1900
57     1904
58     1904
59     1904
60     1904
61     1904
62     1908
63     1908
64     1908
65     1908
66     1912
67     1912
68     1912
69     1912
70     1912
71     1916
72     1916
73     1916
74     1916
75     1920
76     1920
77     1920
78     1920
79     1920
80     1924
81     1924
82     1924
83  

If we want a data frame instead and don't want to use to_frame, we can provide a **list** containing the column name.

In [74]:
# Answer Here don't want to use to_frame,
col = [1, 2, 3, 4, 5]
col = ['A', 'B', 'C', 'D', 'E']
columns = ['Column1_Name', 'Column2_Name']

dataf = pd.DataFrame(list(zip(col, col)), columns=columns)

print(dataf)


  Column1_Name Column2_Name
0            A            A
1            B            B
2            C            C
3            D            D
4            E            E


If we give only one row but many column labels, we'll get back a **Series** corresponding to a row of the table. This new Series has a neat index, where **each entry is the name of the column** that the data came from.

In [37]:
# Answer Here
elections.loc[2] 

Year                      1828
Candidate       Andrew Jackson
Party               Democratic
Popular vote            642806
Result                     win
%                    56.203927
Name: 2, dtype: object

In [ ]:
# Answer Here

If we omit the column argument altogether, the **default behavior is to retrieve all columns**.

In [40]:
# Answer Here
elections.loc[[0,3,6,9]]


Year               Candidate                  Party  Popular vote Result  \
0  1824          Andrew Jackson  Democratic-Republican        151271   loss   
3  1828       John Quincy Adams    National Republican        500897   loss   
6  1832            William Wirt           Anti-Masonic        100715   loss   
9  1836  William Henry Harrison                   Whig        550816   loss   

           %  
0  57.210122  
3  43.796073  
6   7.821583  
9  37.721543

Specify Rows and Columns as List to retrive specific enteries

In [42]:
# Answer Here
elections.loc[[2,4],['Candidate','Year']]

Candidate  Year
2  Andrew Jackson  1828
4  Andrew Jackson  1832

Boolean Series are also boolean arrays, so we can use the Boolean Array Selection from earlier using loc as well.

In [47]:
# Answer Here
election_loss = elections['Result'] == 'loss'
elections.loc[election_loss]

Year               Candidate                  Party  Popular vote Result  \
0    1824          Andrew Jackson  Democratic-Republican        151271   loss   
3    1828       John Quincy Adams    National Republican        500897   loss   
5    1832              Henry Clay    National Republican        484205   loss   
6    1832            William Wirt           Anti-Masonic        100715   loss   
7    1836       Hugh Lawson White                   Whig        146109   loss   
9    1836  William Henry Harrison                   Whig        550816   loss   
10   1840        Martin Van Buren             Democratic       1128854   loss   
12   1844              Henry Clay                   Whig       1300004   loss   
14   1848              Lewis Cass             Democratic       1223460   loss   
15   1848        Martin Van Buren              Free Soil        291501   loss   
18   1852            John P. Hale              Free Soil        155210   loss   
19   1852          Winfield Scott                   Whig       1386942   loss   
21   1856         John C. Frémont             Republican       1342345   loss   
22   1856        Millard Fillmore               American        873053   loss   
24   1860               John Bell   Constitutional Union        590901   loss   
25   1860    John C. Breckinridge    Southern Democratic        848019   loss   
26   1860      Stephen A. Douglas    Northern Democratic       1380202   loss   
28   1864     George B. McClellan             Democratic       1812807   loss   
29   1868         Horatio Seymour             Democratic       2708744   loss   
31   1872          Horace Greeley     Liberal Republican       2834761   loss   
34   1876        Samuel J. Tilden             Democratic       4288546   loss   
35   1880         James B. Weaver              Greenback        308649   loss   
37   1880  Winfield Scott Hancock             Democratic       4444976   loss   
38   1884         Benjamin Butler          Anti-Monopoly        134294   loss   
40   1884         James G. Blaine             Republican       4856905   loss   
41   1884           John St. John            Prohibition        147482   loss   
42   1888          Alson Streeter            Union Labor        146602   loss   
44   1888         Clinton B. Fisk            Prohibition        249819   loss   
45   1888        Grover Cleveland             Democratic       5534488   loss   
46   1892       Benjamin Harrison             Republican       5176108   loss   
48   1892         James B. Weaver               Populist       1041028   loss   
49   1892            John Bidwell            Prohibition        270879   loss   
50   1896          John M. Palmer    National Democratic        134645   loss   
51   1896         Joshua Levering            Prohibition        131312   loss   
52   1896  William Jennings Bryan             Democratic       6509052   loss   
54   1900         John G. Woolley            Prohibition        210864   loss   
55   1900  William Jennings Bryan             Democratic       6370932   loss   
57   1904         Alton B. Parker             Democratic       5083880   loss   
58   1904          Eugene V. Debs              Socialist        402810   loss   
59   1904        Silas C. Swallow            Prohibition        259102   loss   
61   1904        Thomas E. Watson               Populist        114070   loss   
62   1908          Eugene V. Debs              Socialist        420852   loss   
63   1908        Eugene W. Chafin            Prohibition        254087   loss   
64   1908  William Jennings Bryan             Democratic       6408979   loss   
66   1912          Eugene V. Debs              Socialist        901551   loss   
67   1912        Eugene W. Chafin            Prohibition        208156   loss   
68   1912      Theodore Roosevelt            Progressive       4122721   loss   
69   1912            William Taft             Republican       3486242   loss   
71   1916         Allan L. Benson   

## String-labeled Rows

Let's do a quick example using data with string-labeled rows instead of integer labeled rows, just to make sure we're really understanding loc.

Use mottos.csv file

In [49]:
# Answer Here
mottos = pd.read_csv('C:\\Users\\zafar\\Downloads\\mottos.csv')
mottos

State                                              Motto  \
0          Alabama                      Audemus jura nostra defendere   
1           Alaska                                North to the future   
2          Arizona                                         Ditat Deus   
3         Arkansas                                     Regnat populus   
4       California                                    Eureka (Εὕρηκα)   
5         Colorado                                    Nil sine numine   
6      Connecticut                            Qui transtulit sustinet   
7         Delaware                           Liberty and Independence   
8          Florida                                    In God We Trust   
9          Georgia                        Wisdom, Justice, Moderation   
10          Hawaii                  Ua mau ke ea o ka ʻāina i ka pono   
11           Idaho                                      Esto perpetua   
12        Illinois                  State sovereignty, national union   
13         Indiana                          The Crossroads of America   
14            Iowa  Our liberties we prize and our rights we will ...   
15          Kansas                                Ad astra per aspera   
16        Kentucky                               Deo gratiam habeamus   
17       Louisiana                         Union, justice, confidence   
18           Maine                                             Dirigo   
19        Maryland                       Fatti maschii, parole femine   
20   Massachusetts          Ense petit placidam sub libertate quietem   
21        Michigan          Si quaeris peninsulam amoenam circumspice   
22       Minnesota                                   L'étoile du Nord   
23     Mississippi                                   Virtute et armis   
24        Missouri                      Salus populi suprema lex esto   
25         Montana                                        Oro y plata   
26        Nebraska                            Equality before the law   
27          Nevada                                All For Our Country   
28   New Hampshire                                   Live Free or Die   
29      New Jersey                             Liberty and prosperity   
30      New Mexico                                      Crescit eundo   
31        New York                                          Excelsior   
32  North Carolina                                   Esse quam videri   
33    North Dakota                    \nSerit ut alteri saeclo prosit   
34            Ohio                  With God, all things are possible   
35        Oklahoma                                 Labor omnia vincit   
36          Oregon                                Alis volat propriis   
37    Pennsylvania                  Virtue, liberty, and independence   
38    Rhode Island                                               Hope   
39  South Carolina          Dum spiro spero \nAnimis opibusque parati   
40    South Dakota                          Under God the people rule   
41       Tennessee                           Agriculture and Commerce   
42           Texas                                         Friendship   
43            Utah                                           Industry   
44         Vermont                       Stella quarta decima fulgeat   
45        Virginia                                Sic semper tyrannis   
46      Washington                                      Al-ki or Alki   
47   West Virginia                              Montani semper liberi   
48       Wisconsin                                            Forward   
49         Wyoming                                       Equal Rights   

                                          Translation        Language  \
0                          We dare defend our rights!           Latin   
1                                                   —         English   
2                                        God enriches           Latin 

In [56]:
mottos.set_index("State")

Motto  \
State                                                               
Alabama                             Audemus jura nostra defendere   
Alaska                                        North to the future   
Arizona                                                Ditat Deus   
Arkansas                                           Regnat populus   
California                                        Eureka (Εὕρηκα)   
Colorado                                          Nil sine numine   
Connecticut                               Qui transtulit sustinet   
Delaware                                 Liberty and Independence   
Florida                                           In God We Trust   
Georgia                               Wisdom, Justice, Moderation   
Hawaii                          Ua mau ke ea o ka ʻāina i ka pono   
Idaho                                               Esto perpetua   
Illinois                        State sovereignty, national union   
Indiana                                 The Crossroads of America   
Iowa            Our liberties we prize and our rights we will ...   
Kansas                                        Ad astra per aspera   
Kentucky                                     Deo gratiam habeamus   
Louisiana                              Union, justice, confidence   
Maine                                                      Dirigo   
Maryland                             Fatti maschii, parole femine   
Massachusetts           Ense petit placidam sub libertate quietem   
Michigan                Si quaeris peninsulam amoenam circumspice   
Minnesota                                        L'étoile du Nord   
Mississippi                                      Virtute et armis   
Missouri                            Salus populi suprema lex esto   
Montana                                               Oro y plata   
Nebraska                                  Equality before the law   
Nevada                                        All For Our Country   
New Hampshire                                    Live Free or Die   
New Jersey                                 Liberty and prosperity   
New Mexico                                          Crescit eundo   
New York                                                Excelsior   
North Carolina                                   Esse quam videri   
North Dakota                      \nSerit ut alteri saeclo prosit   
Ohio                            With God, all things are possible   
Oklahoma                                       Labor omnia vincit   
Oregon                                        Alis volat propriis   
Pennsylvania                    Virtue, liberty, and independence   
Rhode Island                                                 Hope   
South Carolina          Dum spiro spero \nAnimis opibusque parati   
South Dakota                            Under God the people rule   
Tennessee                                Agriculture and Commerce   
Texas                                                  Friendship   
Utah                                                     Industry   
Vermont                              Stella quarta decima fulgeat   
Virginia                                      Sic semper tyrannis   
Washington                                          Al-ki or Alki   
West Virginia                               Montani semper liberi   
Wisconsin                                                 Forward   
Wyoming                                              Equal Rights   

                                                      Translation  \
State                                                               
Alabama                                We dare defend our rights!   
Alaska                                                          —   
Arizona                                              God enriches   
Arkansas                                          The people rule   
California                                        I have 

Extract slice, can be specified using slice notation, even if the rows have string labels instead of integer labels.

In [58]:
mottos.loc['Alabama':'West Virginia']

Empty DataFrame
Columns: [State, Motto, Translation, Language, Date Adopted]
Index: []

### iloc

loc's cousin iloc is very similar, but is used to access based on numerical position instead of label. For example, to access to the top 3 rows and top 3 columns of a table, we can use [0:3, 0:3]. 'iloc' slicing is **exclusive**, just like standard Python slicing of numerical values.

Use iloc to extract first 3 rows and columns from elections DataFrame

In [55]:
#Answer Here
elections.iloc[0:3,0:3]

Year          Candidate                  Party
0  1824     Andrew Jackson  Democratic-Republican
1  1824  John Quincy Adams  Democratic-Republican
2  1828     Andrew Jackson             Democratic

We will use both `loc` and `iloc` in the course. `loc` is generally preferred for a number of reasons, for example:

1. It is harder to make mistakes since you have to literally write out what you want to get.
2. Code is easier to read, because the reader doesn't have to know e.g., what column #17 represents.
3. It is robust against permutations of the data, e.g. the social security administration switches the order of two columns.

However, iloc is sometimes more convenient. We'll provide examples of when iloc is the superior choice.

## Handy Properties and Utility Functions for Series and DataFrames

The head, shape, size, and describe methods can be used to quickly get a good sense of the data we're working with. For example:

In [61]:
#mottos = pd.read_csv("/content/drive/MyDrive/mottos.csv")
mottos = pd.read_csv('C:\\Users\\zafar\\Downloads\\mottos.csv')
mottos.head()

State                          Motto                 Translation  \
0     Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1      Alaska            North to the future                           —   
2     Arizona                     Ditat Deus                God enriches   
3    Arkansas                 Regnat populus             The people rule   
4  California                Eureka (Εὕρηκα)             I have found it   

  Language Date Adopted  
0    Latin         1923  
1  English         1967  
2    Latin         1863  
3    Latin         1907  
4    Greek         1849

In [ ]:
# Answer Here

Size of DataFrame

In [62]:
# Answer Here
mottos.size

250

The fact that the size is 250 means our data file is relatively small, with only 250 total entries.

Shape of DataFrame

In [63]:
# Answer Here
mottos.shape

(50, 5)

Use describe function and extract the meaningful information from DataFrame

In [64]:
# Answer Here
mottos.describe()

State                          Motto Translation Language  \
count        50                             50          49       50   
unique       50                             50          30        8   
top     Alabama  Audemus jura nostra defendere           —    Latin   
freq          1                              1          20       23   

       Date Adopted  
count            50  
unique           47  
top            1893  
freq              2

Above, we see a quick summary of all the data. For example, the most common language for mottos is Latin, which covers 23 different states. Does anything else seem surprising?

We can get a direct reference to the index using .index.

In [65]:
# Answer Here
mottos.index

RangeIndex(start=0, stop=50, step=1)

In [66]:
mottos.head(2)

State                          Motto                 Translation  \
0  Alabama  Audemus jura nostra defendere  We dare defend our rights!   
1   Alaska            North to the future                           —   

  Language Date Adopted  
0    Latin         1923  
1  English         1967

It turns out the columns also have an Index. We can access this index by using `.columns`.

In [67]:
# Answer Here
mottos.columns

Index(['State', 'Motto', 'Translation', 'Language', 'Date Adopted'], dtype='object')

## Sorting and Value Counts

There are also a ton of useful utility methods we can use with Data Frames and Series. For example, we can create a copy of a data frame sorted by a specific column using `sort_values`.

In [68]:
# Answer Here
ele = elections.copy()
ele.sort_values('%')

Year               Candidate                  Party  Popular vote Result  \
156  2004              David Cobb                  Green        119859   loss   
141  1992                Bo Gritz               Populist        106152   loss   
160  2004        Michael Peroutka           Constitution        143630   loss   
148  1996            John Hagelin            Natural Law        113670   loss   
165  2008        Cynthia McKinney                  Green        161797   loss   
172  2016          Darrell Castle           Constitution        203091   loss   
164  2008           Chuck Baldwin           Constitution        199750   loss   
110  1956      T. Coleman Andrews         States' Rights        107929   loss   
147  1996         Howard Phillips              Taxpayers        184656   loss   
126  1976      Thomas J. Anderson               American        158271   loss   
124  1976           Lester Maddox   American Independent        170274   loss   
125  1976          Roger MacBride            Libertarian        172557   loss   
99   1948        Claude A. Watson            Prohibition        103708   loss   
107  1952        Vincent Hallinan            Progressive        140746   loss   
95   1940           Norman Thomas              Socialist        116599   loss   
136  1988           Lenora Fulani           New Alliance        217221   loss   
132  1984          David Bergland            Libertarian        228111   loss   
181  2020          Howard Hawkins                  Green        405035   loss   
89   1932       William Z. Foster              Communist        103307   loss   
127  1980          Barry Commoner               Citizens        233052   loss   
139  1992            Andre Marrou            Libertarian        290087   loss   
102  1948           Norman Thomas              Socialist        139569   loss   
159  2004        Michael Badnarik            Libertarian        397265   loss   
170  2012              Jill Stein                  Green        469627   loss   
153  2000            Harry Browne            Libertarian        384431   loss   
161  2004             Ralph Nader            Independent        465151   loss   
163  2008                Bob Barr            Libertarian        523715   loss   
92   1936           Norman Thomas              Socialist        187910   loss   
154  2000            Pat Buchanan                 Reform        448895   loss   
138  1988                Ron Paul            Libertarian        431750   loss   
146  1996            Harry Browne            Libertarian        485759   loss   
174  2016           Evan McMullin            Independent        732273   loss   
167  2008             Ralph Nader            Independent        739034   loss   
75   1920        Aaron S. Watkins            Prohibition        188787   loss   
149  1996             Ralph Nader                  Green        685297   loss   
85   1928           Norman Thomas              Socialist        267478   loss   
61   1904        Thomas E. Watson               Populist        114070   loss   
121  1976         Eugene McCarthy            Independent        740460   loss   
51   1896         Joshua Levering            Prohibition        131312   loss   
50   1896          John M. Palmer    National Democratic        134645   loss   
169  2012            Gary Johnson            Libertarian       1275971   loss   
78   1920   Parley P. Christensen           Farmer–Labor        265398   loss   
128  1980                Ed Clark            Libertarian        921128   loss   
177  2016              Jill Stein                  Green       1457226   loss   
180  2020            Jo Jorgensen            Libertarian       1865724   loss   
73   1916             Frank Hanly            Prohibition        221302   loss   
42   1888          Alson Streeter            Union Labor        146602   loss   
38   1884         Benjamin Butler          Anti-Monopoly        134294   loss   
67   1912        Eugene W. Chafin   

As mentioned before, all Data Frame methods return a copy and do **not** modify the original data structure, unless you set inplace to True.

If we want to sort in reverse order, we can set `ascending=False`.

In [69]:
ele.sort_values('%', ascending=False)

Year               Candidate                  Party  Popular vote Result  \
114  1964          Lyndon Johnson             Democratic      43127041    win   
91   1936      Franklin Roosevelt             Democratic      27752648    win   
120  1972           Richard Nixon             Republican      47168710    win   
79   1920          Warren Harding             Republican      16144093    win   
133  1984           Ronald Reagan             Republican      54455472    win   
84   1928          Herbert Hoover             Republican      21427123    win   
86   1932      Franklin Roosevelt             Democratic      22821277    win   
109  1956       Dwight Eisenhower             Republican      35579180    win   
0    1824          Andrew Jackson  Democratic-Republican        151271   loss   
60   1904      Theodore Roosevelt             Republican       7630557    win   
2    1828          Andrew Jackson             Democratic        642806    win   
32   1872           Ulysses Grant             Republican       3597439    win   
106  1952       Dwight Eisenhower             Republican      34075529    win   
27   1864         Abraham Lincoln         National Union       2211317    win   
94   1940      Franklin Roosevelt             Democratic      27313945    win   
4    1832          Andrew Jackson             Democratic        702735    win   
80   1924         Calvin Coolidge             Republican      15723789    win   
97   1944      Franklin Roosevelt             Democratic      25612916    win   
135  1988       George H. W. Bush             Republican      48886597    win   
11   1840  William Henry Harrison                   Whig       1275583    win   
162  2008            Barack Obama             Democratic      69498516    win   
30   1868           Ulysses Grant             Republican       3013790    win   
56   1900        William McKinley             Republican       7228864    win   
8    1836        Martin Van Buren             Democratic        763291    win   
65   1908            William Taft             Republican       7678335    win   
34   1876        Samuel J. Tilden             Democratic       4288546   loss   
178  2020            Joseph Biden             Democratic      81268924    win   
168  2012            Barack Obama             Democratic      65915795    win   
53   1896        William McKinley             Republican       7112138    win   
17   1852         Franklin Pierce             Democratic       1605943    win   
131  1980           Ronald Reagan             Republican      43903230    win   
157  2004          George W. Bush             Republican      62040610    win   
13   1844              James Polk             Democratic       1339570    win   
123  1976            Jimmy Carter             Democratic      40831881    win   
111  1960            John Kennedy             Democratic      34220984    win   
112  1960           Richard Nixon             Republican      34108157   loss   
100  1948            Harry Truman             Democratic      24179347    win   
74   1916          Woodrow Wilson             Democratic       9126868    win   
144  1996            Bill Clinton             Democratic      47400125    win   
12   1844              Henry Clay                   Whig       1300004   loss   
39   1884        Grover Cleveland             Democratic       4914482    win   
45   1888        Grover Cleveland             Democratic       5534488   loss   
176  2016         Hillary Clinton             Democratic      65853514   loss   
151  2000                 Al Gore             Democratic      50999897   loss   
33   1876        Rutherford Hayes             Republican       4034142    win   
36   1880          James Garfield             Republican       4453337    win   
40   1884         James G. Blaine             Republican       4856905   loss   
158  2004              John Kerry             Democratic      59028444   loss   
37   1880  Winfield Scott Hancock   

We can also use `sort_values` on Series objects.

In [70]:
mottos['Language'].sort_values().head(50)

46    Chinook Jargon
49           English
29           English
28           English
27           English
26           English
48           English
37           English
38           English
40           English
17           English
34           English
42           English
14           English
41           English
12           English
1            English
13           English
8            English
7            English
9            English
43           English
22            French
4              Greek
10          Hawaiian
19           Italian
39             Latin
44             Latin
36             Latin
45             Latin
47             Latin
35             Latin
33             Latin
0              Latin
31             Latin
30             Latin
23             Latin
21             Latin
20             Latin
18             Latin
16             Latin
15             Latin
11             Latin
6              Latin
5              Latin
3              Latin
2              Latin
32           

For Series, the `value_counts` method is often quite handy.

In [71]:
mottos['Language'].value_counts()

Latin             23
English           21
Greek              1
Hawaiian           1
Italian            1
French             1
Spanish            1
Chinook Jargon     1
Name: Language, dtype: int64

Also commonly used is the `unique` method, which returns **all unique values** as a numpy array.

In [72]:
mottos['Language'].unique()

array(['Latin', 'English', 'Greek', 'Hawaiian', 'Italian', 'French',
       'Spanish', 'Chinook Jargon'], dtype=object)

In [2]:
def fiba(n):
    if n < 2:
        return n
    else:
        return fiba(n-1) + fiba(n-2)



fiba(5)

5

# Thank you!

In [3]:
#cloumn index integer by col and row 
#iloc for integer value pass col and row  also upper limit not including
 


